types/v2/function.go

```diff
@@ -0,0 +1,92 @@
+// Copyright 2019 OpenFaaS Authors
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+package v2
+
+// Function describes an OpenFaaS function
+type Function struct {
+	// Service is the name of the function
+	Service string `json:"service"`
+
+	// ServiceAccount represents the identity of the function for the orchestration layer
+	ServiceAccount string `json:"serviceAccount,omitempty"`
+
+	// Image corresponds to a container image
+	Image string `json:"image"`
+
+	// Network is specific to Docker Swarm - default overlay network is: func_functions
+	Network string `json:"network"`
+
+	// EnvProcess corresponds to the fprocess variable for your container watchdog.
+	EnvProcess string `json:"envProcess"`
+
+	// EnvVars provides overrides for functions.
+	EnvVars map[string]string `json:"envVars"`
+
+	// RegistryAuth is the registry authentication (optional)
+	// in the same encoded format as Docker native credentials
+	// (see ~/.docker/config.json)
+	RegistryAuth string `json:"registryAuth,omitempty"`
+
+	// Constraints are specific to back-end orchestration platform
+	Constraints []string `json:"constraints"`
+
+	// Secrets list of secrets to be made available to function
+	Secrets []string `json:"secrets"`
+
+	// Labels are metadata for functions which may be used by the
+	// back-end for making scheduling or routing decisions
+	Labels *map[string]string `json:"labels"`
+
+	// Annotations are metadata for functions which may be used by the
+	// back-end for management, orchestration, events and build tasks
+	Annotations *map[string]string `json:"annotations"`
```

Comment:
Maps are already pointers and accept nil, can we accept a plain map here? EnvVars is a plain map.

Reply:
I wonder if it's like this for a reason and what the blast radius would be by making that change?

Reply:
It should be zero blast radius since maps already behave like ptrs and anyone using this will actively be transitioning to this new type. The change wouldn't sneak in on people.

I think the subbed of example ptrs for ptra in the k8s code base come from code generation versus a decision to use ptrs to otrst

Reply:
What about all the code that uses them as pointers and dereferences them?

Reply:
Since this is a new type, none of that code exists yet

Reply:
I would also gladly help to refactor code, if it gets rid of unnecessary pointers

Reply:
This is in every provider, the CLI, the gateway, all the event connectors.

Whilst I appreciate the zeal for Go purity, this change could be done after the structural change without so much re-work / impact across dozens of Git repos.

Maybe I'm wrong? Am I missing something?

Reply:
The current structure can be serialized/deserialized into/from `CreateFunctionRequest` so we can upgrade to v2 gradually. If we change the field types then we have to upgrade all providers and gateway at the same time and break backwards compatibility.